subsys/net/lib/mqtt_socket/mqtt.c

```diff
@@ -0,0 +1,699 @@
+#include "mqtt_internal.h"
+#include "mqtt_os.h"
+
+/** Number of buffers needed per client - one for RX and one for TX */
+#define BUFFER_COUNT_PER_CLIENT 2
+
+/** Total number of buffers needed by the module. */
+#define TOTAL_BUFFER_COUNT (BUFFER_COUNT_PER_CLIENT * MQTT_MAX_CLIENTS)
+
+/** MQTT Client table. */
+static struct mqtt_client *mqtt_client[MQTT_MAX_CLIENTS];
+
+/** Mutex used by MQTT implementation. */
+struct k_mutex mqtt_mutex;
+
+/** Memory slab submitted to the MQTT memory management.*/
+struct k_mem_slab mqtt_slab;
+static char __aligned(4) mqtt_slab_buffer[TOTAL_BUFFER_COUNT *
+					  MQTT_MAX_PACKET_LENGTH];
+
+static u32_t get_client_index(struct mqtt_client *const client)
+{
+	for (u32_t index = 0; index < MQTT_MAX_CLIENTS; index++) {
+		if (mqtt_client[index] == client) {
+			return index;
+		}
+	}
+
+	return MQTT_MAX_CLIENTS;
+}
+
+static void client_free(struct mqtt_client *const client)
+{
+	MQTT_STATE_INIT(client);
+
+	/* Free memory used for TX packets and reset the pointer. */
+	if (client->tx_buf != NULL) {
+		mqtt_free(client->tx_buf);
+		client->tx_buf = NULL;
+	}
+
+	if (client->rx_buf != NULL) {
+		mqtt_free(client->rx_buf);
+		client->rx_buf = NULL;
+	}
+}
+
+static void client_init(struct mqtt_client *const client)
+{
+	memset(client, 0, sizeof(*client));
+
+	MQTT_STATE_INIT(client);
+
+	client->protocol_version = MQTT_VERSION_3_1_0;
+	client->clean_session = 1;
+
+	/* Allocate buffer packets in TX and RX path. */
+	client->tx_buf = mqtt_malloc(MQTT_MAX_PACKET_LENGTH);
+	client->rx_buf = mqtt_malloc(MQTT_MAX_PACKET_LENGTH);
+}
+
+/**@brief Notifies event to the application.
+ *
+ * @param[in] client Identifies the client for which the procedure is requested.
+ * @param[in] evt Reason for disconnection.
+ */
+void event_notify(struct mqtt_client *const client, const struct mqtt_evt *evt,
+		  u32_t flags)
+{
+	const mqtt_evt_cb_t evt_cb = client->evt_cb;
+
+	if (evt_cb != NULL) {
+		mqtt_mutex_unlock();
+
+		evt_cb(client, evt);
+
+		mqtt_mutex_lock();
+
+		if (flags & MQTT_EVT_FLAG_INSTANCE_RESET) {
+			client_free(client);
+			client_init(client);
+		}
+	}
+}
+
+/**@brief Notifies disconnection event to the application.
+ *
+ * @param[in] client Identifies the client for which the procedure is requested.
+ * @param[in] result Reason for disconnection.
+ */
+void disconnect_event_notify(struct mqtt_client *client, int result)
+{
+	const u32_t client_index = get_client_index(client);
+	struct mqtt_evt evt;
+
+	/* Remove the client from internal table. */
+	if (client_index != MQTT_MAX_CLIENTS) {
+		mqtt_client[client_index] = NULL;
+	}
+
+	/* Determine appropriate event to generate. */
+	if (MQTT_VERIFY_STATE(client, MQTT_STATE_CONNECTED) ||
+	    MQTT_VERIFY_STATE(client, MQTT_STATE_DISCONNECTING)) {
+		evt.type = MQTT_EVT_DISCONNECT;
+		evt.result = result;
+	} else {
+		evt.type = MQTT_EVT_CONNACK;
+		evt.result = -ECONNREFUSED;
+	}
+
+	/* Free the instance. */
+	client_free(client);
```

Comment:
Both looks weird to have it before event_notify(client, ...) and not needed, as you pass  MQTT_EVT_FLAG_INSTANCE_RESET reset there (though is that flag is again needed? why not do it straight and free client after event delivery?).